testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -9314,11 +9314,12 @@ private static void checkGroupConcatFuncFails(SqlOperatorFixture t) {
   }
 
   private static void checkArrayAggFunc(SqlOperatorFixture f) {
-    f.setFor(SqlLibraryOperators.ARRAY_CONCAT_AGG, VM_FENNEL, VM_JAVA);
+    f.setFor(SqlLibraryOperators.ARRAY_AGG, VM_FENNEL, VM_JAVA);
```

Comment:
When fixing some similar bugs I have found that a RelOptRulesTest can often exhibit the bug by producing a wrong plan.
Here is an example: https://github.com/apache/calcite/pull/3289/files#diff-a430972be1169f680b7886b824720254fa1dabe37191db34bd2a63f10cf7a438
I don't know if it will work in your case as well.

Reply:
Thanks @mihaibudiu.
The main issue with this question is incorrect function return type inference, and it seem's not be related to the planner rules. And the old test case of this line uses the wrong operator.

Reply:
Sorry @mihaibudiu, I read the example again and it makes sense to me.
I can add some test in `RelOptRulesTest` as well.

Reply:
But I can't find any proper rules to verify that `ARRAY_AGG($0)` should returns  `ARRAY NULL` type.

Reply:
The optimization will replace the plan with a query returning constant null.

Reply:
Maybe I'm wrong but I'm not sure `RelOptRulesTest` is the best place to be testing appropriate return types. If you want to see the rel plan generated by certain sql, maybe `SqlToRelConverterTest` could be more what you are looking for?

Reply:
I have found that sometimes optimization rules produce results that differ from the runtime results for the same expression. In particular `CoreRules.PROJECT_REDUCE_EXPRESSIONS` will optimize constant expressions. Currently there is no easy way to check that the optimization produces the expected result. Does the SqlToRelConverterTest allow you to specify optimizations?

I am working towards https://issues.apache.org/jira/browse/CALCITE-5891, once that exists all the existing SqlOperatorTests will be automatically run in both ways. (But it may take awhile, it has uncovered at least 50 bugs in RelToSql and the optimizer.)

Reply:
This was just a suggestion, I don't think it is required necessarily for this PR.